gobblin-core/src/main/java/gobblin/source/DatePartitionedSource.java

```diff
@@ -0,0 +1,279 @@
+   * @param lowWaterMark
+   * @param fileCount
+   * @param maxFilesPerJob
+   * @param tableType
+   * @param multiWorkUnitWeightedQueue
+   */
+  private void addNewWorkUnits(SourceState state, DateTime lowWaterMark, int fileCount, int maxFilesPerJob,
+      TableType tableType, MultiWorkUnitWeightedQueue multiWorkUnitWeightedQueue) {
+
+    Path sourceDir = new Path(state.getProp(ConfigurationKeys.SOURCE_FILEBASED_DATA_DIRECTORY));
+    DateTime currentDay = new DateTime();
+
+    // Process all data from the lowWaterMark date until the maxFilesPerJob has been hit
+    for (DateTime date = lowWaterMark; !date.isAfter(currentDay) && fileCount < maxFilesPerJob; date = date.plusDays(1)) {
+
+      String topicName = sourceDir.getName();
+
+      Path dayPath =
+          new Path(sourceDir, DEFAULT_DATE_PARTITIONED_SOURCE_DAILY_FOLDER_NAME + Path.SEPARATOR
+              + DAILY_FOLDER_FORMATTER.print(date));
+
+      try {
+        if (this.fs.exists(dayPath)) {
+
+          // Create an extract object for the dayPath
+          SourceState partitionState = new SourceState();
+
+          partitionState.addAll(state);
+          partitionState.setProp(ConfigurationKeys.SOURCE_ENTITY, topicName);
+
+          Extract extract =
+              partitionState.createExtract(tableType,
+                  partitionState.getProp(ConfigurationKeys.EXTRACT_NAMESPACE_NAME_KEY), topicName);
+
+          LOG.info("Created extract: " + extract.getExtractId() + " for path " + dayPath);
+
+          // Create a workunit for each file in the folder
+          for (FileStatus fileStatus : this.fs.listStatus(dayPath, getFileFilter())) {
+
+            LOG.info("Will process file " + fileStatus.getPath());
+
+            partitionState.setProp(ConfigurationKeys.SOURCE_FILEBASED_FILES_TO_PULL, fileStatus.getPath());
+            partitionState.setProp(ConfigurationKeys.WORK_UNIT_LOW_WATER_MARK_KEY, WATER_MARK_FORMATTER.print(date));
+            partitionState.setProp(ConfigurationKeys.WORK_UNIT_HIGH_WATER_MARK_KEY, WATER_MARK_FORMATTER.print(date));
+
+            WorkUnit singleWorkUnit = partitionState.createWorkUnit(extract);
+
+            multiWorkUnitWeightedQueue.addWorkUnit(singleWorkUnit, fileStatus.getLen());
+
+            fileCount++;
+          }
+        } else {
+          LOG.info("Path " + dayPath + " does not exist, skipping");
+        }
+      } catch (IOException e) {
+        Throwables.propagate(e);
+      }
+    }
+
+    LOG.info("Total number of files extracted for the current run: " + fileCount);
+  }
+
+  /**
+   * Gets the LWM for this job runs; calculates it by looking at the work unit states from the previous job. It also
+   * implements the wrap around logic, if the new LWM hits the wrap around date then it get set to its default value.
+   * @param state the source state
+   */
+  private DateTime getLowWaterMark(SourceState state) {
+    long lwm = state.getPropAsLong(DATE_PARTITIONED_SOURCE_MIN_WATERMARK_VALUE);
```

Comment:
Set all the configs in the init() method.

I put the default value to 0, so by default the job will start consuming data from Jan 1st, 1970. I changed the MIN_WATERMARK to be stored as a unix timestamp, so that if someone wants to change to ddMMyyyy they only have to change the Joda Formatters and the documentation.